modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy

```diff
@@ -284,8 +288,18 @@ class FilePorter {
                 log.debug "Local cache found for foreign file ${source.toUriString()} at ${target.toUriString()}"
                 return target
             }
+
             log.debug "Copying foreign file ${source.toUriString()} to work dir: ${target.toUriString()}"
-            return FileHelper.copyPath(source, target)
+
+            // copy to a temp file first and move atomically to ensure that a partial copy doesn't get cached 
+            // as completed.
+            // use the target parent dir as the temp dir because atomic moves between partitions are not supported,
+            // and /tmp is usually a different partition.
+            Path tmpFile = Files.createTempFile(target.getParent(), target.getFileName().toString(), null)
```

Comment:
This is the best way for a local file system, but I fear the implementation of `createTempFile` would mess up things when using a remote storage such S3, etc. 

A better approach could be to use a UUID random string as a temporary file name eg 

```
def temp = target.getParent().resolve( UUID.randomUUID().toString() )
FileHelper.copyPath(source, temp)
FileHelper.movePath(temp, target)
```

For the same reason better to avoid `ATOMIC_MOVE` option which may not be supported by all file system supported by NF.

Reply:
Updated. The tmp-fs implementation worries me a bit - this sums it up: https://rachelbythebay.com/w/2020/08/11/files/

Also, I didn't actually see my debug log weirdly but from what I can tell it's working.

Reply:
Nice! but if you didn't see the code very likely it hasn't be executed 😅 . Have tested using 

```
make compile 
./launch.sh run <usual command line>
```

Reply:
Ah, okay, got it.

```
Aug-21 01:11:54.972 [FileTransfer-thread-1] DEBUG nextflow.file.FilePorter - Copying foreign file s3://josh-test-e6fbd807-96e6-4626-962f-1cf84ad7aa9e/myfile.txt to work dir: /home/j/third-party/nextflow/work/stage/69/61efbef013a67070686a7b8a16874f/myfile.txt
Aug-21 01:11:54.972 [FileTransfer-thread-1] DEBUG nextflow.file.FilePorter - Using temp file /home/j/third-party/nextflow/work/stage/69/61efbef013a67070686a7b8a16874f/temp-c3634d82-aae3-4c37-b5a7-897659fe1b73 for /home/j/third-party/nextflow/work/stage/69/61efbef013a67070686a7b8a16874f/myfile.txt
```

```
root@606b59bab953:/home/j/third-party/nextflow# cat work/b3/1c5b83bcdb9e2a1523e54c90ba20fb/.command.out 
hello world
```

Reply:
Well done!